### PR TITLE
[Bugfix][ROCm][MoE] Fix mxfp4 oracle regressions from #37128

### PR DESCRIPTION
**Pull request by AndreasKaratzas**
Fixes several issues introduced by #37128 that broke gpt-oss on ROCm.

- Restore gfx950 gate for CK mxfp4 backend selection. The old code only picked CK on gfx950 via on_gfx950(), the refactor dropped this and let CK get selected on gfx942 where it crashes.
- Restore CK_MXFP4_MOE_DIM_ALIGNMENT (256) check. Models with intermediate_size not aligned to 256 (like gpt-oss-20b at 2880) hit a reshape error in aiter shuffle_scale_a16w4. Added is_supported_config to AiterExperts so the backend selector falls through to Triton.
- Restore hidden_pad/intermediate_pad for CK path. These were passed to rocm_aiter_ops.fused_moe() in the old code but got lost in the refactor. Added fields to FusedMoEQuantConfig and wired them through.

Tested on MI325X (gfx942):
- test_gpt_oss_speculative_reasoning_leakage passes
- GPQA eval via gpt_oss.evals: 56.76% (1584 questions, effort=low)
- Backend correctly falls back to Triton on non-gfx950

Related:
- https://github.com/vllm-project/vllm/pull/37811
- https://github.com/vllm-project/vllm/pull/37784
- https://github.com/vllm-project/vllm/pull/37786
- https://github.com/vllm-project/vllm/pull/37128

cc @kenroche

### CODE REVIEW

File: vllm/lora/layers/fused_moe.py [hunk 1]
```diff
@@ -109,8 +109,8 @@ def _get_lora_moe_configs(
         else:  # fall back to the default config
             get_config_func = functools.partial(
                 try_get_optimal_moe_lora_config,
-                w1_shape=layer.w13_weight.size(),
-                w2_shape=layer.w2_weight.size(),
+                w1_shape=layer.w13_weight.shape,
```
Comment (jeejeelee) [inline vllm/lora/layers/fused_moe.py:112]: Do these changes make sense?

File: vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py [hunk 2]
```diff
@@ -257,7 +257,7 @@ def triton_kernel_moe_forward(
         # sparse_logits.indx contains global expert IDs – remap to local.
         topk_ids = expert_map[sparse_logits.indx.to(torch.long)]
         topk_weights = sparse_logits.vals
-        local_num_experts = w1.size(0)
+        local_num_experts = w1.shape[0]
```
Comment (jeejeelee) [inline vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py:260]: Do these changes make sense?

File: vllm/lora/ops/triton_ops/fused_moe_lora_op.py [hunk 3]
```diff
@@ -354,7 +354,7 @@ def _fused_moe_lora_kernel(
         cur_k_offset = k * (BLOCK_SIZE_K * SPLIT_K)
         k_remaining = K - cur_k_offset
         # pre-fetch lora weight
-        if b_desc is not None:
+        if USE_TMA and b_desc is not None:
```
Comment (jeejeelee) [inline vllm/lora/ops/triton_ops/fused_moe_lora_op.py]: I don't think this change is necessary.

File: vllm/lora/ops/triton_ops/fused_moe_lora_op.py [hunk 4]
```diff
@@ -369,7 +369,7 @@ def _fused_moe_lora_kernel(
                 b = tl.load(b_ptrs, mask=b_mask, other=0.0)
             b_ptrs += BLOCK_SIZE_K * SPLIT_K * stride_bk
 
-        if a_desc is not None:
+        if USE_TMA and a_desc is not None:
```
Comment (jeejeelee) [inline vllm/lora/ops/triton_ops/fused_moe_lora_op.py]: DITTO

File: vllm/lora/ops/triton_ops/fused_moe_lora_op.py [hunk 5]
```diff
@@ -379,7 +379,11 @@ def _fused_moe_lora_kernel(
             )
             a_ptrs += BLOCK_SIZE_K * SPLIT_K * stride_ak
 
-        accumulator += tl.dot(a, b)
+        # Cast operands to matching dtype for tl.dot. On ROCm, Triton's
+        # compiler may infer different types for a and b when merging
+        # if/else branches (TMA desc path returns fp32, tl.load returns
+        # the pointer's element type).
+        accumulator += tl.dot(a.to(tl.bfloat16), b.to(tl.bfloat16))
```
Comment (jeejeelee) [inline vllm/lora/ops/triton_ops/fused_moe_lora_op.py:386]: On the NV platform, will this introduce overhead?

File: vllm/lora/ops/triton_ops/fused_moe_lora_op.py (hunk 3 above)
Comment (AndreasKaratzas) [reply vllm/lora/ops/triton_ops/fused_moe_lora_op.py]: Indeed. Reverting

File: vllm/lora/ops/triton_ops/fused_moe_lora_op.py (hunk 4 above)
Comment (AndreasKaratzas) [reply vllm/lora/ops/triton_ops/fused_moe_lora_op.py]: reverting

File: vllm/lora/layers/fused_moe.py (hunk 1 above)
Comment (AndreasKaratzas) [reply vllm/lora/layers/fused_moe.py:112]: When mxfp4 quantization is active, w1/w2 can be triton_kernels.tensor.Tensor objects, not torch.Tensor. Calling .size(0) or .dim() on them would raise AttributeError. The change to .shape[0] and len(.shape) makes the code work with both tensor types.

File: vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py (hunk 2 above)
Comment (AndreasKaratzas) [reply vllm/model_executor/layers/fused_moe/gpt_oss_triton_kernels_moe.py:260]: (same 😅)

File: vllm/lora/ops/triton_ops/fused_moe_lora_op.py (hunk 5 above)
Comment (AndreasKaratzas) [reply vllm/lora/ops/triton_ops/fused_moe_lora_op.py:386]: On NVIDIA with TMA enabled, both a and b come from TMA descriptor loads which already return fp32. The .to(tl.bfloat16) adds a cast, but this is a register-level type conversion so it's essentially free compared to the tl.dot itself. Triton's compiler would already need to reconcile types for the dot product.

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py [hunk 6]
```diff
@@ -292,6 +296,8 @@ def rocm_aiter_fused_experts(
             doweight_stage1=apply_router_weight_on_input,
             num_local_tokens=num_local_tokens,
             output_dtype=output_dtype,
+            hidden_pad=quant_config.hidden_pad // 128 * 128,
```
Comment (Rohan138) [inline vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: I think this should just be `quant_config.hidden_pad`? I.e. don't div by 128 * 128: https://github.com/ROCm/aiter/blob/main/aiter/fused_moe.py#L1018

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py [hunk 7]
```diff
@@ -349,6 +355,38 @@ def _supports_parallel_config(moe_parallel_config: FusedMoEParallelConfig) -> bo
             or moe_parallel_config.use_fi_nvl_one_sided_kernels
         )
 
+    @staticmethod
```
Comment (Rohan138) [inline vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: Currently `is_supported_config` is called by `select_mxfp4_moe_backend` before we pad the hidden and intermediate sizes in `mxfp4_round_up_hidden_size_and_intermediate_size`: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/mxfp4.py#L101, so this will fully disable CKMOE for gpt-oss.

File: vllm/model_executor/layers/quantization/mxfp4.py [hunk 8]
```diff
@@ -163,6 +163,12 @@ def create_weights(
             intermediate_size_per_partition_after_pad
         )
 
+        # CK (gfx950) padding info for rocm_aiter_ops.fused_moe()
+        self.hidden_pad = extra_weight_attrs.get("hidden_pad", 0)
+        self.intermediate_pad = (
```
Comment (Rohan138) [inline vllm/model_executor/layers/quantization/mxfp4.py:163]: This will work currently because the layer is padding the intermediate_sizes, but it's a bit of a hack. https://github.com/vllm-project/vllm/pull/34285 will address this more cleanly once it is rebased.

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py (hunk 6 above)
Comment (AndreasKaratzas) [reply vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: Done :)

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py (hunk 7 above)
Comment (AndreasKaratzas) [reply vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: Done :)

File: vllm/model_executor/layers/quantization/mxfp4.py (hunk 8 above)
Comment (AndreasKaratzas) [reply vllm/model_executor/layers/quantization/mxfp4.py:163]: Done :)

File: vllm/lora/layers/fused_moe.py (hunk 1 above)
Comment (BowenBao) [reply vllm/lora/layers/fused_moe.py:112]: why did this work prior to https://github.com/vllm-project/vllm/pull/37128? do you know what was changed that lead to this?

Comment (AndreasKaratzas) [reply vllm/lora/layers/fused_moe.py:112]: Before #37128, layer.w13_weight was always a torch.Tensor (a torch.nn.Parameter), so .size() worked fine.

#37128 introduced the Mxfp4MoEMethod.process_weights_after_loading path which calls convert_to_mxfp4_moe_kernel_format and _swizzle_mxfp4. For TRITON backends, this converts the weights to triton_kernels.tensor.Tensor and assigns them directly to layer.w13_weight / layer.w2_weight

triton_kernels.tensor.Tensor has .shape but not .size() or .dim()

The same applies to the .dim() in to len(.shape) and .size(0) in to .shape[0] changes in gpt_oss_triton_kernels_moe.py and modular_kernel.py.

**Comment by AndreasKaratzas** [comment]
> Please share if there's a discussion with upstream folks. #37128 broke CI, and rather than reverting it to figure out a proper solution, there's pressure to quickly land this PR. The size -> shape change and padding logic feel like quick(hacky) band-aids that will require the ROCm team to do more follow-up work down the line.
> 
> I don't feel this aligns with OSS best practices. However, if the maintainers are aware of this and have agreed it's the best way forward, I have nothing to add. cc @tjtanaa, @gshtras, @dllehr-amd, @ChuanLi1101

The .size() in to .shape change isn't a band-aid, it's the correct fix. triton_kernels.tensor.Tensor (used for MXFP4 swizzled weights since #37128) exposes .shape but not .size() or .dim(). `.shape` is the common interface that works with both tensor types and is the more Pythonic/numpy-standard accessor anyway. There's no follow-up work needed here.

On the padding logic, I hear the concern about it being incremental. The CK MXFP4 kernels on gfx950 require 256-byte aligned dimensions, and the padding was already happening in create_weights but wasn't being properly communicated to the layer/config.

File: vllm/lora/layers/fused_moe.py (hunk 1 above)
Comment (BowenBao) [reply vllm/lora/layers/fused_moe.py:112]: thanks for detailed explanation. would it be possible to instead extract the underlying torch.Tensor from the triton wrapped tensors
```
  # After _swizzle_mxfp4 returns wrapped tensors:
  w13_weight = w13_weight.storage.data  # Extract torch.Tensor
  w2_weight = w2_weight.storage.data
```
I assume the layout/swizzle information is already preserved separately in PrecisionConfig. I hope this makes it easier for downstream code, outside of triton kernel, to remain torch.Tensor API compatible.

Comment (AndreasKaratzas) [reply vllm/lora/layers/fused_moe.py:112]: The forward path needs both w1.storage.data (the raw packed torch.Tensor, passed to moe_gemm_a8w4) and w1.shape (the logical tensor shape <E, N, K> used for assertions and size extraction). These are different because the raw tensor has different dimensions due to FP4 packing (2 values per byte) and swizzle layout.

If we stored only .storage.data on the layer, the kernel code at gpt_oss_triton_kernels_moe.py:463 would work, but shape checks like E, _, N = w1.shape would give wrong values (packed dimensions, not logical).

The triton_kernels.tensor.Tensor wrapper exists precisely to carry both the raw data and the logical shape/layout metadata together. Using .shape is the correct way I think to interact with it.

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py (hunk 7 above)
Comment (Rohan138) [reply vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: if you just need to check for the current device being gfx950, you should instead override `supports_current_device` instead of overriding the top-level `is_supported_config`: https://github.com/vllm-project/vllm/blob/8de5261e69af81aaf7f68d420b69199653cd1882/vllm/model_executor/layers/fused_moe/modular_kernel.py#L570

Comment (AndreasKaratzas) [reply vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: > if you just need to check for the current device being gfx950, you should instead override `supports_current_device` instead of overriding the top-level `is_supported_config`:
> 
> https://github.com/vllm-project/vllm/blob/8de5261e69af81aaf7f68d420b69199653cd1882/vllm/model_executor/layers/fused_moe/modular_kernel.py#L570

_supports_current_device has no visibility into the quant scheme. It would reject AiterExperts entirely on non-gfx950. The gfx950 requirement is specific to the MXFP4 weight type, so I modified _supports_quant_scheme (which receives weight_key). Also removed is_supported_config override.

File: vllm/model_executor/layers/fused_moe/oracle/mxfp4.py [hunk 9]
```diff
@@ -212,7 +212,15 @@ def select_mxfp4_moe_backend(
...
     if config.is_lora_enabled:
         if not current_platform.is_cuda():
-            raise NotImplementedError("Mxfp4 LoRA only supported on CUDA Platform.")
+            # ROCm: Triton mxfp4 LoRA hits GPU memory faults due to
+            # triton_kernels.tensor.Tensor / HIP read-only page issues
+            # during weight swizzle and LoRA forward. Needs work from
+            # the triton_kernels/aiter side. This matches pre-#37128.
+            raise NotImplementedError(
+                "Mxfp4 LoRA is only supported on CUDA. "
+                "ROCm support is blocked by triton_kernels.tensor.Tensor "
```
Comment (gshtras) [inline vllm/model_executor/layers/fused_moe/oracle/mxfp4.py]: Nit. not is_cuda doesn't automatically mean rocm. Other platforms may get confused by this message.

Comment (AndreasKaratzas) [reply vllm/model_executor/layers/fused_moe/oracle/mxfp4.py]: Done :)

**Review by Rohan138 — APPROVED** [review]

**Review by gshtras — APPROVED** [review]

File: vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py [hunk 10]
```diff
@@ -332,7 +334,15 @@ def _supports_quant_scheme(
             (kFp8StaticChannelSym, kFp8DynamicTokenSym),
             (kMxfp4Static, None),
         ]
-        return (weight_key, activation_key) in SUPPORTED_W_A
+        if (weight_key, activation_key) not in SUPPORTED_W_A:
+            return False
+        # CK MXFP4 MoE kernels are only supported on gfx950.
+        if weight_key == kMxfp4Static:
+            from vllm.platforms.rocm import on_gfx950
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/rocm_aiter_fused_moe.py]: Aren't there side effects from importing `vllm.platforms.rocm` directly?
